src/Qir/Runtime/lib/QIR/context.cpp

```diff
@@ -9,58 +9,90 @@
+    QirExecutionContext::QirExecutionContext(IRuntimeDriver* simulator, bool trackAllocatedObjects)
+        : driver(simulator)
+        , trackAllocatedObjects(trackAllocatedObjects)
     {
-        assert(g_context == nullptr);
-        g_context = std::make_unique<QirExecutionContext>(sim, trackAllocatedObjects);
-
-        if (g_context->driver != nullptr)
+        if (this->trackAllocatedObjects)
         {
-            ResultOne = g_context->driver->UseOne();
-            ResultZero = g_context->driver->UseZero();
+            this->allocationsTracker = std::make_unique<AllocationsTracker>();
         }
-        else
+    }
+
+    // If we just remove this user-declared-and-defined dtor 
+    // then it will be automatically defined together with the class definition,
+    // which will require the `AllocationsTracker` to be a complete type 
+    // everywhere where `public/QirContext.hpp` is included 
+    // (we'll have to move `allocationsTracker.hpp` to `public/`).
+    QirExecutionContext::~QirExecutionContext() = default;
+
+    void QirExecutionContext::Init(IRuntimeDriver* simulator, bool trackAllocatedObjects /*= false*/)
+    {
+        assert(GlobalContext() == nullptr);
+        GlobalContext() = std::make_unique<QirExecutionContext>(simulator, trackAllocatedObjects);
+    }
+
+    void QirExecutionContext::Deinit()
+    {
+        assert(GlobalContext() != nullptr);
+
+        if (GlobalContext()->trackAllocatedObjects)
         {
-            ResultOne = nullptr;
-            ResultZero = nullptr;
+            GlobalContext()->allocationsTracker->CheckForLeaks();
         }
+
+        GlobalContext().reset(nullptr);
     }
 
-    void ReleaseQirContext()
+    void QirExecutionContext::OnAddRef(void* object)
     {
-        assert(g_context != nullptr);
-
-        if (g_context->trackAllocatedObjects)
+        if(trackAllocatedObjects)
         {
-            g_context->allocationsTracker->CheckForLeaks();
+            this->allocationsTracker->OnAddRef(object);
         }
+    }
 
-        ResultOne = nullptr;
-        ResultZero = nullptr;
-        g_context.reset(nullptr);
+    void QirExecutionContext::OnRelease(void* object)
+    {
+        if(this->trackAllocatedObjects)
+        {
+            this->allocationsTracker->OnRelease(object);
+        }
     }
 
-    QirExecutionContext::QirExecutionContext(IRuntimeDriver* sim, bool trackAllocatedObjects)
-        : driver(sim)
-        , trackAllocatedObjects(trackAllocatedObjects)
+    void QirExecutionContext::OnAllocate(void* object)
     {
-        if (this->trackAllocatedObjects)
+        if(this->trackAllocatedObjects)
         {
-            this->allocationsTracker = std::make_unique<AllocationsTracker>();
+            this->allocationsTracker->OnAllocate(object);
         }
     }
-    QirExecutionContext::~QirExecutionContext() = default;
+
+    IRuntimeDriver* QirExecutionContext::GetDriver() const

```

Comment:
@swernli, any comments about why the simulator is called Driver?
I believe it has been renamed to Driver during the recent mass-refactoring couple of weeks ago.

Reply:
This terminology is something I switched to because this interface is not really exclusive for the simulator or for host program, but rather the object that allows for invoking functions on the backend target. That might be a simulator, a direct link to hardware, or some other intermediate component that forwards the invocations to an entirely different infrastructure. In this configuration, what was the host program is really expected to be in the QIR itself (or perhaps some thin logic in the `main` function the QIR is compiled with).